### PR TITLE
Apply normalization callback when updating data with CLI input

### PR DESCRIPTION
**Pull request by happz**
Before running a plugin, its data package is updated with CLI options, if there were any given by user. Normalization callback must be applied to values tmt extracts from Click options - it is not enough to use normalization callbacks as values for Click's `type=` parameter, because Click treats them in different way, and expects different results. This has unfortunate effects especially when `field(multiple=True)` was used.

Instead, we need to apply them explicitely when moving data from CLI to plugin data packages, but we do have a code ready just for that, therefore after a small refactoring, users of `field()` should be covered.

### CODE REVIEW

**Comment by happz** [comment]
@psss adding this to 1.21 milestone, as a prerequisite of https://github.com/teemtee/tmt/pull/1752 and the future development of the area.

File: tmt/utils.py [hunk 1]
```diff
@@ -3607,6 +3604,62 @@ def __init__(
...
+
+    else:
+        value = raw_value
+
+    # Set attribute by adding it to __dict__ directly. Messing with setattr()
+    # might cause re-use of mutable values by other instances.
+    container.__dict__[keyname] = value
+
+    # TODO: keep this code for debugging purpose - when logging changes with
+    # https://github.com/teemtee/tmt/pull/1685, this would be turned into
```
Comment (psss) [inline tmt/utils.py:3642]: Logging changes in #1685 have been already merged. Shall we update this part to use proper logging calls as suggested? Or keep it as is to enable only when really needed?

File: tmt/utils.py [hunk 2]
```diff
@@ -3706,7 +3759,14 @@ def _normalize_string_list(
...
 
-        return [value] if isinstance(value, str) else value
+        if isinstance(value, str):
+            return [value]
+
+        if isinstance(value, (list, tuple)):
+            return [item for item in value]
+
+        raise SpecificationError(
+            f"Field can be either string or list of strings, '{type(value)}' found.")
```
Comment (psss) [inline tmt/utils.py:3769]: Could we modify the normalization calls to accept the `keyname` as well? It would allow to mention the key in the erorr message. Currently we get:

```
Failed to load step data for PrepareInstallData: Field can be either string or list of strings, '<class 'bool'>' found.
The exception was caused by the previous exception:
Field can be either string or list of strings, '<class 'bool'>' found.
```

From this message user cannot see which field has incorrect value. I guess this would probably be material for a separate pull request.

**Review by psss — CHANGES_REQUESTED** [review]
Looks nice, thanks! Just a few minor comments.

File: tmt/utils.py (hunk 2 above)
Comment (happz) [reply tmt/utils.py:3769]: I believe we can, good point. I'd do that in a standalone PR.

File: tmt/steps/__init__.py [hunk 3]
```diff
@@ -904,11 +904,20 @@ def enabled_on_guest(self, guest: 'Guest') -> bool:
             return True
         return where in (guest.name, guest.role)
 
-    def _update_data_from_options(self, keys: List[str]) -> None:
-        for key in keys:
-            value = self.opt(tmt.utils.key_to_option(key))
-            if value:
-                setattr(self.data, key, value)
+    def _update_data_from_options(self, keys: Optional[List[str]] = None) -> None:
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/steps/__init__.py:907]: Added in cce47a92ffa216c736d25acd9066d8e6ba52e5ea.

File: tmt/utils.py (hunk 1 above)
Comment (happz) [reply tmt/utils.py]: Changed in cce47a92ffa216c736d25acd9066d8e6ba52e5ea - logging is enabled, but with a deep debug level.

File: tests/unit/test_dataclasses.py [hunk 4]
```diff
@@ -1,8 +1,114 @@
...
+
+    dataclass_normalize_field(data, 'foo', 2, root_logger)
+    assert data.foo == 2
+
+    dataclass_normalize_field(data, 'foo', '3', root_logger)
+    assert data.foo == 3
+
+    with pytest.raises(
+            tmt.utils.SpecificationError,
+            match=r"Field 'foo' can be either unset or integer, '<class 'str'>' found."):
```
[the remark this answers is not in the document]
Comment (happz) [reply tests/unit/test_dataclasses.py]: Changed slightly in cce47a92ffa216c736d25acd9066d8e6ba52e5ea, although I'm not happy about it. I'd rather add an exception for this, or modify the `SpecificationError`, to accept keyname & found type and render the message for us. Right now the message is generated in several places, hard to keep in sync. Another PR?

[the remark this answers is not in the document]
Comment (psss) [reply tests/unit/test_dataclasses.py]: Thanks for the adjustment, looks better. Yes, let's handle the further improvements (modifying `SpecificationError` sounds good to me) in a separate pull request.

File: tmt/utils.py (hunk 2 above)
Comment (psss) [reply tmt/utils.py]: Ack.

File: tmt/utils.py (hunk 1 above)
Comment (psss) [reply tmt/utils.py]: Nice, thanks!

**Review by psss — APPROVED** [review]
Looks good, thanks for the adjustments.
